numba/typed/dictobject.py

```diff
@@ -1055,24 +1057,252 @@ def impl_iterator_iternext(context, builder, sig, args, result):
+            items.append(pair)
+        ret = cgctx.make_tuple(builder, sig.return_type, items)
+        return ret
+    kvs = [types.Tuple((types.unicode_type, x)) for x in d.types]
+    sig = types.Tuple(kvs)(d)
+    return sig, impl
+
+
+@overload_method(types.LiteralStrKeyDict, 'items')
+def literalstrkeydict_impl_items(d):
+    if not isinstance(d, types.LiteralStrKeyDict):
+        return
+
+    def impl(d):
+        return _str_items_mixed_values_to_tuple(d)
+    return impl
+
+
+@overload(operator.contains)
+def literalstrkeydict_impl_contains(d, k):
+    if not isinstance(d, types.LiteralStrKeyDict):
+        return
+
+    def impl(d, k):
+        for key in d.keys():
+            if k == key:
+                return True
+        return False
+    return impl
+
+
+@overload(operator.setitem)
+def literalstrkeydict_banned_impl_setitem(d, key, value):
+    if not isinstance(d, types.LiteralStrKeyDict):
+        return
+    raise TypingError("Cannot mutate a literal dictionary")
+
+
+@overload(operator.delitem)
+def literalstrkeydict_banned_impl_delitem(d, k):
+    if not isinstance(d, types.LiteralStrKeyDict):
+        return
+    raise TypingError("Cannot mutate a literal dictionary")
+
+
+@overload_method(types.LiteralStrKeyDict, 'popitem')
+@overload_method(types.LiteralStrKeyDict, 'pop')
+@overload_method(types.LiteralStrKeyDict, 'clear')
+@overload_method(types.LiteralStrKeyDict, 'setdefault')
+def literalstrkeydict_banned_impl_mutators(d, *args):
+    if not isinstance(d, types.LiteralStrKeyDict):
+        return
+    raise TypingError("Cannot mutate a literal dictionary")
+
+
+@lower_cast(types.LiteralStrKeyDict, types.LiteralStrKeyDict)
+def cast_LiteralStrKeyDict_LiteralStrKeyDict(context, builder, fromty, toty,
+                                             val):
+    # should have been picked up by typing
+    for (k1, v1), (k2, v2) in zip(fromty.literal_value.items(),
+                                  toty.literal_value.items()):
+        # these checks are just guards, typing should have picked up any
+        # problems
+        if k1 != k2: # keys must be same
+            msg = "LiteralDictionary keys are not the same {} != {}"
+            raise LoweringError(msg.format(k1, k2))
+        # values must be same ty
+        if context.typing_context.unify_pairs(v1, v2) is None:
+            msg = "LiteralDictionary values cannot by unified, have {} and {}"
+            raise LoweringError(msg.format(v1, v2))
+    else:
+        fromty = types.Tuple(fromty.types)
+        toty = types.Tuple(toty.types)
+        olditems = cgutils.unpack_tuple(builder, val, len(fromty))
+        items = [context.cast(builder, v, f, t)
+                 for v, f, t in zip(olditems, fromty, toty)]
+        return context.make_tuple(builder, toty, items)
+
+
+@lower_cast(types.DictType, types.DictType)
+def cast_DictType_DictType(context, builder, fromty, toty, val):
+    # should have been picked up by typing
+    return val
```

Comment:
untested

Reply:
how do we need this?

Reply:
think it's left over from trying to make literal dictionary types gracefully degrade, removed in 120ffce5f

Reply:
Turns out it is needed: https://dev.azure.com/numba/numba/_build/results?buildId=6009&view=logs&j=12ecc1d6-2964-5850-b419-89f950957853&t=abce8f77-76e0-5291-fb7e-5ca9f390cd04&l=9914 IIRC `typed.Dict` uses the list dtype and not its initial value, but the initial value is part of the type, so casts such as the one failing won't work without it.

Reply:
Reverted 120ffce in ad7343f

Reply:
`Cannot cast DictType[int32,int32]<iv={4: 2}> to DictType[int32,int32]<iv={1: 3}>:`

why is that allowed? should it be unified to one without initial_value?

Reply:
Think this stems from two problems:
1. No attempt at unification in the cast impl https://github.com/numba/numba/blob/e043ee705d65abd3a066500be4d62b54bdf3586c/numba/typed/typedobjectutils.py#L23-L35
2. `DictType` will only unify with another `DictType` if the other is "imprecise" https://github.com/numba/numba/blob/e043ee705d65abd3a066500be4d62b54bdf3586c/numba/core/types/containers.py#L636-L643